### PR TITLE
fix(security): replace raw SQL where parameter with parameterized filters

### PR DESCRIPTION
**Pull request by CybotTM**
## Summary

The ReadTable tool's `where` parameter accepted arbitrary SQL fragments with only a 7-keyword blacklist (DROP, DELETE, UPDATE, INSERT, TRUNCATE, ALTER, CREATE). This was trivially bypassable via UNION SELECT, subqueries, CASE/WHEN, and function calls — enabling data exfiltration from any table accessible to the backend user.

## Changes

- **Replace `where` with `filters`**: Structured array of `{field, operator, value}` objects using parameterized queries exclusively
- **Field validation**: Against TCA columns and essential fields, with access control enforcement via `canAccessField()`
- **Operator whitelist**: 12 safe comparison operators (eq, neq, lt, lte, gt, gte, like, notLike, in, notIn, isNull, isNotNull)
- **Parameterized values**: All values through `createNamedParameter()` with proper ParameterType
- **Value validation**: Comparison operators require a non-null value; in/notIn require arrays
- **Test runner**: Docker-based `Build/runTests.sh` matching CI environment (PHP 8.3, SQLite)

## Breaking change

`where` parameter replaced by `filters` — per project rules, MCP tool parameters don't need backwards compatibility. Old `where` parameter is silently ignored.

## Test plan

- [x] All 441 functional tests pass
- [x] 26 new security-focused tests covering all operators, validation, and injection rejection
- [x] Existing tests migrated from `where` to `filters`

## Summary by CodeRabbit

* **New Features**
  * Added structured table read filtering via a `filters` array (`{field, operator, value}`), supporting equality/inequality, `like/notLike`, range comparisons, `in/notIn`, and null checks.
  * Multiple filters are combined with AND and applied consistently to both returned records and the total count.
* **Bug Fixes**
  * Legacy raw `where` input is no longer supported; invalid/unsafe filter definitions now return clear errors.
* **Tests**
  * Updated and expanded functional coverage for schema exposure, operator/value validation, malformed inputs, security edge cases, and large filter handling.

### CODE REVIEW

**Comment by CybotTM** [comment]
Any further work required, or any questions?

**Comment by Nemo64** [comment]
I've started work on a way to automatically evaluate different LLMs with the MCP tools.
I am currently a bit hesitant to import any changes to tools, since they always have unpredictable behavioural changes and manual testing is very anecdotal and time consuming.

The implementation itself looks good; I just haven't had time to really test it, sorry.

File: Classes/MCP/Tool/Record/ReadTableTool.php [hunk 1]
```diff
@@ -143,7 +161,7 @@ protected function doExecute(array $params): CallToolResult
                 fn($n) => $n > 0
             ));
         }
-        $condition = $params['where'] ?? '';
+        $filters = $params['filters'] ?? [];
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/MCP/Tool/Record/ReadTableTool.php:164]: Addressed in fcb9360 — a non-array `filters` value is now rejected up front with a `ValidationException` ("The "filters" parameter must be an array of filter objects"), so a scalar can no longer reach the typed `array $filters` argument and raise a `TypeError`. Covered by the new `testRejectsNonArrayFiltersParameter`.

File: Classes/MCP/Tool/Record/ReadTableTool.php [hunk 2]
```diff
@@ -680,6 +682,141 @@ protected function normalizeFieldNames(string $table, array $requestedFields): a
...
+                        throw new ValidationException(["Filter at index {$index}: '{$operator}' operator requires an array value"]);
+                    }
+                    $arrayType = $this->isIntegerArray($value)
+                        ? \TYPO3\CMS\Core\Database\Connection::PARAM_INT_ARRAY
+                        : \TYPO3\CMS\Core\Database\Connection::PARAM_STR_ARRAY;
+                    $exprMethod = $operator === 'in' ? 'in' : 'notIn';
+                    $queryBuilder->andWhere($expr->$exprMethod(
+                        $resolvedField,
+                        $queryBuilder->createNamedParameter($value, $arrayType)
+                    ));
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Classes/MCP/Tool/Record/ReadTableTool.php:818]: Addressed in fcb9360 — `applyFilters()` now rejects non-scalar values for the scalar operators (eq/neq/lt/lte/gt/gte/like/notLike) and requires `in`/`notIn` to be a non-empty, flat, scalar array. Malformed payloads now fail as a clean `ValidationException` instead of a runtime DBAL binding error. Covered by `testRejectsScalarOperatorWithArrayValue`, `testRejectsEmptyInArray`, and `testRejectsNestedInArray`.

File: Tests/Functional/MCP/Tool/ReadTableFilterSecurityTest.php [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+        ]);
+
+        $this->assertFalse($result->isError, json_encode($result->jsonSerialize()));
+        $data = $this->extractJsonFromResult($result);
+
+        $this->assertGreaterThan(0, count($data['records']));
+        foreach ($data['records'] as $record) {
+            $this->assertStringContainsString('Welcome', $record['header']);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Functional/MCP/Tool/ReadTableFilterSecurityTest.php:107]: Addressed in fcb9360 — the LIKE/notLike assertions now use `assertStringContainsStringIgnoringCase` / `assertStringNotContainsStringIgnoringCase` to match the case-insensitive collation.

File: Tests/Functional/MCP/Tool/ReadTableToolTest.php [hunk 4]
```diff
@@ -207,42 +207,54 @@ public function testMissingTableParameter(): void
...
+
+        $this->assertFalse($result->isError, json_encode($result->jsonSerialize()));
         $data = json_decode($result->content[0]->text, true);
-        
+
         // All returned records should have CType = textmedia
         foreach ($data['records'] as $record) {
             $this->assertEquals('textmedia', $record['CType']);
         }
     }
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Functional/MCP/Tool/ReadTableToolTest.php:232]: Addressed in fcb9360 — added `$this->assertGreaterThan(0, count($data['records']));` before the loop so it can't pass vacuously. Applied the same guard to the equivalent filter tests in `ReadTableFilterSecurityTest`.

**Comment by CybotTM** [comment]
Addressed the review feedback in fcb9360 (amended into the single commit, CI green across TYPO3 13.4/14.3 × PHP 8.2–8.5).

**Input validation (the two `Major` items):**
- A non-array `filters` parameter is now rejected with a `ValidationException` before it reaches `getRecords()`, instead of letting a scalar hit the typed `array $filters` argument and throw a `TypeError`.
- `applyFilters()` rejects non-scalar values for the scalar operators and requires `in`/`notIn` to be a non-empty, flat, scalar array — so malformed payloads fail as a clean validation error rather than a runtime DBAL binding error.

**Tests:** added coverage for the above; made the previously-vacuous filter tests assert a non-empty result set (and rewrote the `neq` case to filter on `header`, since every fixture row is `textmedia`); the raw-`where` test now compares against an unfiltered baseline; and `LIKE` assertions are case-insensitive to match the collation. The nitpicks (relocating `testInOperatorRequiresArrayValue`, etc.) are included too.

No behavioural change for valid input — these only reject malformed filters earlier and more clearly.

File: Tests/Functional/MCP/Tool/ReadTableFilterSecurityTest.php [hunk 5]
```diff
@@ -0,0 +1,504 @@
...
+                ['field' => 'uid', 'operator' => 'gt', 'value' => 100],
+            ],
+        ]);
+
+        $this->assertFalse($result->isError, json_encode($result->jsonSerialize()));
+        $data = $this->extractJsonFromResult($result);
+
+        foreach ($data['records'] as $record) {
+            $this->assertGreaterThan(100, $record['uid']);
+        }
```
[the remark this answers is not in the document]
Comment (CybotTM) [reply Tests/Functional/MCP/Tool/ReadTableFilterSecurityTest.php:124]: Addressed in b6d0adc — added `$this->assertGreaterThan(0, count($data['records']));` before the loop in `testFilterByGreaterThan`, matching the neighboring operator tests.
